csrc/host_ir/jit.cpp

```diff
@@ -166,29 +178,388 @@ void printLlvmIr(llvm::Function* func, std::string_view msg) {
+  std::vector<Merge*> last_level_merge_ops;
+  
+  // push all logical domains as boundary values
+  for(IterDomain* id : logical_domain) {
+    boundary_vals[id->as<Val>()] = false;
+  }
+
+  // push all allocation domains extents in regular order
+  for (auto [i, id] : enumerate(allocation_domain)) {
+    llvm::Value* extent_value = getOrCreateValueForExtent(id, val_to_value, builder);
+    level_order_domains.pushBack(id, extent_value);
+  }
+
+  // traverse backward from allocation domains to logical domains
+  for (Expr* transform : DependencyCheck::getAllExprsBetween(
+    {logical_domain.begin(), logical_domain.end()},
+    {allocation_domain.begin(), allocation_domain.end()}) | std::views::reverse) {
+    if (auto* split = dynamic_cast<Split*>(transform)) {
+      const auto [outer_extent_value, outer_i] = level_order_domains.erase(split->outer());
+      NVF_ERROR(outer_i == level_order_domains.end() || outer_i->first != split->inner(), split->toString(), " is not a valid split");
+      const auto [inner_extent_value, inner_i] = level_order_domains.erase(split->inner());
+            
+      // NOTE: how do we handle device dimension? Currently we just divide it out in split
+      // However, if both dimension are device dimension, do we need to collapse them?
+      // So that in a merge op, we can merge between two local iter domain between one device dimension?
+      llvm::Value* lhs = outer_extent_value;
+      llvm::Value* rhs = inner_extent_value;
+      if(split->outer()->isDeviceDim()) {
+        lhs = builder.getInt64(1);
+      }
+      if(split->inner()->isDeviceDim()) {
+        rhs = builder.getInt64(1);
+      }
+      
+      llvm::Value* in_extent = builder.CreateMul(lhs, rhs);
+      level_order_domains.insert(inner_i, split->in(), in_extent);
+
+    } else if (auto* merge = dynamic_cast<Merge*>(transform)) {
+      if(mapToInputDomain(merge->out(), boundary_vals)!= nullptr && 
+        mapToInputDomain(merge->inner(), boundary_vals)!= nullptr) {
+        last_level_merge_ops.push_back(merge);
+      }
+      const auto [out_extent_value, out_i] = level_order_domains.erase(merge->out());
+      
+      // NOTE: we don't have a protocol to decide which iter domain to pad,
+      // currently we just pad inner value, so dividend is outer value
+      // so inner_extent = (out_extent + outer_extent - 1) / outer_extent, which is a ceilDiv
+      llvm::Value* outer_extent_value = getOrCreateValueForExtent(merge->outer(), val_to_value, builder);
+      llvm::Value* minus_one = builder.CreateSub(outer_extent_value, builder.getInt64(1));
+      llvm::Value* plus_value = builder.CreateAdd(out_extent_value, minus_one);
+      llvm::Value* inner_extent_value = builder.CreateUDiv(plus_value, outer_extent_value);
+        
+      level_order_domains.insert(out_i, merge->outer(), outer_extent_value);
+      level_order_domains.insert(out_i, merge->inner(), inner_extent_value);
```

Comment:
I thought `level_order_domains` are ordered from inner loops to outer loops. So the insert order should be the opposite?

Reply:
I'm pretty confused about what `outer`/`inner` mean inside the merge op vs in the split op above.